compiler/rustc_infer/src/infer/canonical/canonicalizer.rs

```diff
@@ -126,19 +126,52 @@ impl<'tcx> InferCtxt<'tcx> {
     /// handling of `'static` regions (e.g. trait evaluation).
     pub fn canonicalize_query_keep_static<V>(
         &self,
-        value: V,
+        value: ty::ParamEnvAnd<'tcx, V>,
         query_state: &mut OriginalQueryValues<'tcx>,
-    ) -> Canonical<'tcx, V>
+    ) -> Canonical<'tcx, ty::ParamEnvAnd<'tcx, V>>
     where
         V: TypeFoldable<TyCtxt<'tcx>>,
     {
-        Canonicalizer::canonicalize(
+        self.canonicalize_query_with_mode(
+            value,
+            query_state,
+            &CanonicalizeFreeRegionsOtherThanStatic,
+        )
+    }
+
+    fn canonicalize_query_with_mode<V>(
+        &self,
+        value: ty::ParamEnvAnd<'tcx, V>,
+        query_state: &mut OriginalQueryValues<'tcx>,
+        canonicalize_region_mode: &dyn CanonicalizeMode,
+    ) -> Canonical<'tcx, ty::ParamEnvAnd<'tcx, V>>
+    where
+        V: TypeFoldable<TyCtxt<'tcx>>,
+    {
+        let (param_env, value) = value.into_parts();
+        let base = self.tcx.canonical_param_env_cache.get_or_insert(
+            param_env,
+            query_state,
+            |query_state| {
+                Canonicalizer::canonicalize(
+                    param_env,
+                    self,
+                    self.tcx,
+                    &CanonicalizeFreeRegionsOtherThanStatic,
```

Comment:
This does not spark joy 😁 Does anything break if you change this to `CanonicalizeFreeRegions`? I want to get away from canonicalizing static completely. iirc the last time I tried after #102472 there was still some failure in project or sth 🤔 would prefer for you (or me) to first try to  also canonicalize `'static` again to check why it matters.

Reply:
Borrowck error in [`tests/ui/trivial-bounds/trivial-bounds-object.rs`](https://github.com/rust-lang/rust/blob/5f73b006d2ebdb9fb245d952a40abac4a68ba3ec/tests/ui/trivial-bounds/trivial-bounds-object.rs)

ICE in [`tests/ui/nll/issue-61311-normalize.rs`](https://github.com/rust-lang/rust/blob/5f73b006d2ebdb9fb245d952a40abac4a68ba3ec/tests/ui/nll/issue-61311-normalize.rs) and [`tests/ui/nll/issue-61320-normalize.rs`](https://github.com/rust-lang/rust/blob/5f73b006d2ebdb9fb245d952a40abac4a68ba3ec/tests/ui/nll/issue-61320-normalize.rs). MCVE:
```rust
pub trait HasProj {
    type Proj;
}

impl<T: ?Sized> HasProj for T {
    type Proj = ();
}

fn test()
where
    dyn Send + 'static: HasProj,
    <dyn Send + 'static as HasProj>::Proj: Clone,
{
    let _ = test;
}
```

Reply:
cc #118965. It is also the root cause behind the ICE.